Persist copy_ in training graph for inputs that don't require grad

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111046

In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) 

We keep the input mutation in the graph by wrapping the original callable in a wrapper function where in the end we add input.copy_(updated_input) call which is then traced via make_fx. Previously, this was only enabled for forward-only path but unconditionally disabled for joint graph. 

Another caveat is that when we are tracing through tensor subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code logic quite ugly for no obvious performance improvement. 

Most of the changes in this PR are mechanical and I didn't have to make any change to the partitioner. Previously forward/backward heavily relied on metadata field `num_mutated_inps` to figure out whether something is returned as extra output or not. But now since we keep some mutations in the graph, we need to propogate something similar to `num_mutated_inps - num_graph_handled_inps`. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

tugsbayasgalan committed:
Persist copy_ in training graph for buffers

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Persist copy_ in training graph for buffers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1702,6 +1702,18 @@ def functionalized_f_helper(*args):
                     inpt_old.copy_(inpt_new)
 
+        if aot_config.keep_inference_input_mutations and trace_joint:
+            for i, (inpt_old, inpt_f) in enumerate(zip(args[0], f_args[0])):
```
args[1] will be inputs to backward so we ignore them here

bdhirsh replied:
Instead of these two c onditions:
```
if aot_config.keep_inference_input_mutations and not trace_joint:
    ...
if aot_config.keep_inference_input_mutations and trace_joint:
    ...
```

Can you try to merge them together? (I think you'll just have to do e.g. `args_, f_args_ = args[0], f_args[0] if trace_joint else args, f_args`)

tugsbayasgalan committed:
Update on "[WIP] Persist copy_ in training graph for buffers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang commented:
do you want high level feedback while it's WIP? Need more pr desc

tugsbayasgalan committed:
Update on "[WIP] Persist copy_ in training graph for buffers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Persist copy_ in training graph for buffers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Persist copy_ in training graph for buffers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -2062,6 +2063,197 @@ def f(a):
+        out_test = compiled_f(inp_test.clone())
+
+        eager_version_counters = [buffer._version for _, buffer in model_for_eager.named_buffers()]
+        compile_version_counters = [buffer._version for _, buffer in model_for_compile.named_buffers()]
```
mentioned offline - this test might be easier to do with inductor, since (as we learned haha) aot_eager will double-count version counter bumps (since aot_eager will just run each node under an fx interpreter, including `copy_()`, and autograd will see it).

Maybe we can "fix" this in aot_eager by having that backend run under inference_mode, although that seems like it should be its own separate fix.

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1039,9 +1060,16 @@ def inner(*flat_args):
                 # because they show up in the autograd.Function.forward as outputs
-                input_requires_grad_info.append(
-                    isinstance(f_arg, torch.Tensor) and f_arg.requires_grad
-                )
```
Why did this change?

If this is the thing where "input mutations on subclasses never go in the graph in the training case", I don't think we should change the requires_grad field here. Instead, in `ViewAndMutationMeta.__post_init__()` we should just directly use `self.is_subclass`

tugsbayasgalan replied:
Some inputs don't show up as forward outputs i think.

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -624,9 +637,15 @@ def __post_init__(self):
-        self.num_mutated_data_inputs = len(
-            [x for x in self.input_info if x.mutates_data]
-        )
+        if not self.is_subclass and self.keep_input_mutations and self.is_train:
```
A comment here would be nice describing the situation:

In fact, it would be great to have a larger comment here laying out exactly when input mutations are / are not in the graph in the first place (with "training graph where input mutation happens to a subclass" being one of those cases0.

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -992,6 +1011,8 @@ def run_functionalized_fw_and_collect_metadata(
     keep_input_mutations: bool,
     # TODO: refactor to kill this flag
     is_train: bool = False,
+    num_params_buffers: int = 0,
```
this is unnecessary now right?

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1200,9 +1228,17 @@ def inner(*flat_args):
-        assert len(requires_grad_info) == len(output_info) + len(
-            [x for x in input_info if x.mutates_data or x.mutates_metadata]
-        )
+        # TODO (tmanlaibaatar) figure out why we can't just collapse this branch
```
agreed 😛

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1439,6 +1476,9 @@ def maybe_to_fresh_input(idx, t, meta):
     if idx in meta.mutated_inp_indices:
+        if meta.is_train and meta.keep_input_mutations and not meta.input_info[idx].requires_grad:
+            # TODO (tmanlaibaatar)
+            pass
```
?

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1502,11 +1542,18 @@ def inner_fn(*args):
-            for (i, x) in enumerate(args_maybe_cloned)
-            if meta.input_info[i].mutates_metadata or meta.input_info[i].mutates_data
-        ]
+        if meta.is_train and meta.keep_input_mutations:
```
I think this branch is also unnecessary?

also - this is inside of `fn_prepped_for_autograd`, so we should be guaranteed to be in the training case here

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3406,6 +3530,8 @@ def metadata_fn(*primals):
         metadata_fn,
         keep_input_mutations=meta.keep_input_mutations,
         is_train=meta.is_train,
+        num_params_buffers=aot_config.num_params_buffers,
```
we don't need this anymore?

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3328,6 +3451,7 @@ def aot_dispatch_subclass(
     is_joint_structure: bool,
     meta: ViewAndMutationMeta,
     fw_only: Callable,
+    aot_config: AOTConfig,
```
I think we don't need this anymore either?

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -499,6 +499,15 @@ def __post_init__(self):
+    return input_info.mutates_data and not input_info.mutates_metadata and not m.requires_grad
+
+
+def _is_mutation_reflected_in_the_runtime(input_info: InputAliasInfo):
```
Hmm, why is this function not a simple opposite of `_is_mutation_recorded_in_the_graph` (why can't we just kill this function and do `not _is_mutation_recorded_in_the_graph()`?

tugsbayasgalan replied:
Oh this should have been gone. I never use this

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2056,6 +2114,25 @@ def compute_overlapping_inputs(fwd_inputs, aliased_input_indices):
+    return input_info.mutates_data and not input_info.mutates_metadata and not input_info.requires_grad
+
+
+def _check_if_mutation_will_be_returned_in_runtime(input_info: InputAliasInfo, keep_input_mutations: bool, is_train: bool):
```
It seems like we have a few variants of this function. Can we just have a single function `_is_mutation_recorded_in_the_graph()` that takes in all the necessary args and is the source of truth for whether a given mutation will be in the graph vs. the runtime epilogue?

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2395,7 +2484,15 @@ def create_synthetic_base_metadata(
 
     # grab the original requires grad info on the outputs, except the ones from the mutated inputs
-    num_original_input_data_mutations = len([x for x in m.input_info if x.mutates_data or x.mutates_metadata])
+    # TODO (tmanlaibaatar) figure out why we can't just collapse this branch
```
agreed :)

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2668,6 +2765,7 @@ def wrapped_flat_fn(*args):
             wrapped_flat_fn,
             keep_input_mutations=fw_metadata.keep_input_mutations,
             is_train=fw_metadata.is_train,
+            num_params_buffers=aot_config.num_params_buffers,
```
no longer needed

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2815,11 +2913,13 @@ def wrapped_flat_fn(*args):
             wrapped_flat_fn,
             keep_input_mutations=fw_metadata.keep_input_mutations,
             is_train=fw_metadata.is_train,
+            num_params_buffers=aot_config.num_params_buffers,
```
no longer needed

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2815,11 +2913,13 @@ def wrapped_flat_fn(*args):
-            f'ref_metadata={pprint.pformat(partial_asdict(ref_fw_metadata))}, '
-            f'\nactual_metadata={pprint.pformat(partial_asdict(fw_metadata_updated))}'
-        )
+        if not (fw_metadata.keep_input_mutations and fw_metadata.is_train):
```
🤔

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3698,6 +3824,14 @@ def _compiled_autograd_key(ctx):
             args = deduped_flat_tensor_args
+
+            marked_dirty_inps = []
+            if fw_metadata.keep_input_mutations and fw_metadata.is_train:
```
nit: `fw_metadata.is_train` is guaranteed to be true here, no need to check it (we're inside the `autograd.Function`)

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3698,6 +3824,14 @@ def _compiled_autograd_key(ctx):
+
+            marked_dirty_inps = []
+            if fw_metadata.keep_input_mutations and fw_metadata.is_train:
+                for ix, i in enumerate(deduped_flat_tensor_args):
```
Instead of looping over all `deduped_flat_tensor_args`, can you loop over a list of indices of inputs that need to be marked dirty?

It's pretty common for a compiled function to have e.g. 200+ inputs, where only a handful are mutated, so we want to avoid looping over all the inputs in the hot-path (this function is part of the AOTAutograd runtime hotpath).

bdhirsh reviewed (commented):
high level looks good! The majority of my comments are all around refactoring / testing

tugsbayasgalan committed:
Update on "[WIP] Persist copy_ in training graph for buffers"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) Previously, this was only enable for forward-only path but unconditionally disabled for joint graph. Another caveat is that when we are tracing through subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code quite ugly for no obvious performance improvement. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) Previously, this was only enable for forward-only path but unconditionally disabled for joint graph. Another caveat is that when we are tracing through subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code quite ugly for no obvious performance improvement. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) Previously, this was only enable for forward-only path but unconditionally disabled for joint graph. Another caveat is that when we are tracing through subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code quite ugly for no obvious performance improvement. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "[WIP] Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) Previously, this was only enable for forward-only path but unconditionally disabled for joint graph. Another caveat is that when we are tracing through subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code quite ugly for no obvious performance improvement. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan commented on torch/_functorch/aot_autograd.py:
```diff
@@ -596,19 +597,29 @@ class ViewAndMutationMeta:
     is_train: bool = False
+    # Whether we are tracing through subclasses
+    # (ideally should remove this)
+    is_subclass: bool = False
```
I think this path is triggered only when there is some or all inputs are subclasses. In that case, we do the default thing where we just return the mutated inputs instead of trying to keep input mutations in the graph.

bdhirsh replied:
bump on this rename :)

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -596,19 +597,29 @@ class ViewAndMutationMeta:
         mutated_inp_indices = [
-            i for i, m in enumerate(self.input_info) if m.mutates_metadata or m.mutates_data
+            i for i, m in enumerate(self.input_info)
+            if m.mutates_data or m.mutates_metadata
```
avoid needless reformatting like this, it makes review annoying. If you want to ufmt the file before working on it, do it in a separate PR

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1745,7 +1763,7 @@ def functionalized_f_helper(*args):
             # This allows us to factor aot autograd much more nicely, since only one area of the code needs to worry
             # about synthetic bases.
-            for i, (inpt_old, inpt_f) in enumerate(zip(args, f_args)):
+            for i, (inpt_old, inpt_f) in enumerate(zip(args, f_args) if not trace_joint else zip(args[0], f_args[0])):
```
when trace_joint, args is (primals, tangents), only do primals

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1754,12 +1772,14 @@ def functionalized_f_helper(*args):
-                    if meta.input_info[i].mutations_hidden_from_autograd:
-                        with torch.no_grad(), torch.autograd._unsafe_preserve_version_counter(inpt_old):
+                    if not trace_joint:
+                        assert inpt_new is not inpt_old
```
Tugsuu says that this assert passes now

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1754,12 +1772,14 @@ def functionalized_f_helper(*args):
-                        with torch.no_grad(), torch.autograd._unsafe_preserve_version_counter(inpt_old):
+                    if not trace_joint:
+                        assert inpt_new is not inpt_old
+                    if not trace_joint or (trace_joint and not inpt_old.requires_grad):
```
Should test for user input that requires grad

ezyang replied:
It is OK to test requires_grad here because the inner function is performing autograd, so it must be that we have accurate autograd information on the arguments being fed in there. (If we commuted these two wrappers, this would no longer necessarily be true)

ezyang replied:
From a code structure perspective, I feel it would be better if this conditional was based on information in meta.input_info, rather than directly querying inpt_old, because the decision you make here affects the calling conventions everywhere else, which I assume is the plumbing you're doing...

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1754,12 +1772,14 @@ def functionalized_f_helper(*args):
-                        with torch.no_grad(), torch.autograd._unsafe_preserve_version_counter(inpt_old):
+                    if not trace_joint:
+                        assert inpt_new is not inpt_old
+                    if not trace_joint or (trace_joint and not inpt_old.requires_grad):
```
Add an assert that `not inpt_new.requires_grad`

ezyang replied:
I'm not sure this assert is true

ezyang replied:
So the example I am interested in is this:

```
def f(x, y):
   x.add_(y)
```

where x does not require grad and y requires grad.

On entry to create_functionalized_fn, I expect to get args with requires_grad (False, True). When I am done, I expect f_args for x to now require grad (because I inplace mutated it with a tensor that requires grad.) I am not sure if you can somehow do it soundly, but is certainly safest to NOT persist copy_ in this case.

Now, the potential difficulty in the code below is that you are only testing `inpt_old.requires_grad`. Which I would expect to be False for x, even for the example above. So this seems a bit suspicious.

ezyang replied:
We checked and it turned out this is wrong, but input_info[i].requires_grad has the requires grad info AFTER the function is run and works; but this needs to be documented in the InputInfo

ezyang replied:
comments from discuss on why cloning is necessary

```
def f(x)
  x.mul_(2)
  return x

y = f(x)
torch.autograd.grad(y, x)  # identity (1)

x_new = x.clone()
y = f(x_new)
torch.autograd.grad(y, x)  # 2
```

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2115,6 +2136,31 @@ def compute_overlapping_inputs(fwd_inputs, aliased_input_indices):
+        input_info.requires_grad,
+        keep_input_mutations,
+        is_train
+    )
```
You should be calling this from the tracing code

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2115,6 +2136,31 @@ def compute_overlapping_inputs(fwd_inputs, aliased_input_indices):
     return actual_aliased_indices
 
+
+def _check_can_be_in_graph(mutates_data, mutates_metadata, requires_grad, keep_input_mutations: bool, is_train: bool):
```
docblocks would be helpful

ezyang replied:
rename to `_check_if_mutation_can_be_in_graph`

ezyang replied:
The meaning of `requires_grad` is not here so make sure to xref it (requires grad here refers to the requires gradness AFTER you've run the function)

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2115,6 +2136,31 @@ def compute_overlapping_inputs(fwd_inputs, aliased_input_indices):
 
+
+def _check_can_be_in_graph(mutates_data, mutates_metadata, requires_grad, keep_input_mutations: bool, is_train: bool):
+    if keep_input_mutations and is_train:
```
Why not just say `if not keep_input_mutations: return Fales`

bdhirsh replied:
Can you remind me: why do we want to branch on `is_train` here? It seems to me like even if we have an inference graph, if have an input mutation on an input that requires_grad then we probably shouldn't keep it in the graph.

It occurs to me now that since https://github.com/pytorch/pytorch/pull/111347 landed, it's technically possible for this to happen (we have an input mutation on an input that requires grad, but the graph is an inference graph). But I think this is probably wrong, and we just just keep these input mutations out of the graph anyway.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2115,6 +2136,31 @@ def compute_overlapping_inputs(fwd_inputs, aliased_input_indices):
+
+def _check_if_mutation_will_be_returned_in_runtime(input_info: InputAliasInfo, keep_input_mutations: bool, is_train: bool):
+    return (
+        (input_info.mutates_data or input_info.mutates_metadata) and
```
Why do I need this condition?

ezyang replied:
It's because there's a ternary condition going on. In that case, we should have a single function that returns a three-valued enum (MUTATION_INSIDE_GRAPH, MUTATION_OUTSIDE_GRAPH, NO_MUTATION)

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -596,19 +597,29 @@ class ViewAndMutationMeta:
         ]
-        # pre-compute the indices of the inputs that are mutated.
-        # When keep_input_mutations is set, we don't need to worry about our epilogue
-        # handling data-only mutations, because we keep them directly in the graph.
```
what happened to the comment?

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -596,19 +597,29 @@ class ViewAndMutationMeta:
+            i for i, m in enumerate(self.input_info)
+            if _check_if_mutation_can_be_handled_in_fw_graph(m, self.keep_input_mutations, self.is_train) and not self.is_subclass
         ]
+        self.mutated_graph_handled_indices = mutated_graph_handled_indices
```
I need comments on all of these

bdhirsh replied:
+1 (these new is should also be explicitly declared in the dataclass above)

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -596,19 +597,29 @@ class ViewAndMutationMeta:
+
+        mutated_graph_handled_indices = [
+            i for i, m in enumerate(self.input_info)
+            if _check_if_mutation_can_be_handled_in_fw_graph(m, self.keep_input_mutations, self.is_train) and not self.is_subclass
```
We just established that there is a ternary condition, but actually the indices here are not reflecting the ternary; instead mutated_inp_indices is everything, mutated_inp_runtime_indices is MUTATION_OUTSIDE_GRAPH, and mutated_graph_handled_indices is MUTATION_INSIDE_GRAPH

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -596,19 +597,29 @@ class ViewAndMutationMeta:
         mutated_inp_runtime_indices = [
-            i for i, m in enumerate(self.input_info) if m.mutates_metadata or (not self.keep_input_mutations and m.mutates_data)
+            i for i, m in enumerate(self.input_info)
+            if _check_if_mutation_will_be_returned_in_runtime(m, self.keep_input_mutations, self.is_train) and not self.is_subclass
```
Why isn't self.is_subclass passed into the helper function?

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3850,7 +3935,7 @@ def forward(ctx, *deduped_flat_tensor_args):
                 return_new_outs=False
             )
-            return tuple(raw_returns)
+            return tuple(raw_returns) + tuple(marked_dirty_inps)
```
@albanD How come inputs which are `mark_dirty` have to be explicitly returned as outputs to the autograd function? Seems a little dorky to me

albanD replied:
It is necessary in the case where the input is a view so that we can do the proper graph rewrite.
But I agree that in this case, it does not help. We can skip this check when there are multiple outputs (since if the input was a view, it would have failed before). Do you want me to send a PR?

bdhirsh replied:
In theory, AOTAutograd would still need to handle this even if we skipped the check when there are multiple outputs, right? Since if your compiled function happens to mutate a single input and return nothing, then we would still be forced to do the mark_dirty logic.

So I guess skipping the check in most cases could be nice in that it would help us avoid some runtime overhead (not having to manipulate python lists in the hotpath to filter the mark_dirty tensors in most cases). But we still need some conditional logic in AOTAutograd to "sometimes" return mark-dirty inputs. e.g:

```
@torch.compile
def f(x)
    # x will be marked dirty, and autograd.Function will still require us to return it as an output here
    x.mul_(2)
    return
```

albanD replied:
Yes that is correct.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3812,25 +3891,31 @@ def forward(ctx, *deduped_flat_tensor_args):
             if CompiledFunction.metadata.num_unsafe_view_outputs > 0:
                 for idx in CompiledFunction.metadata.unsafe_view_out_indices:
-                    raw_return_idx = num_mutated_inputs + idx
+                    raw_return_idx = num_mutated_runtime_inps + idx
```
so you could have done this in a prep PR

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3865,27 +3950,33 @@ def backward(ctx, *flat_args):
+            if num_graph_handled_inputs > 0:
+                flat_args = flat_args[:-num_graph_handled_inputs]
+            else:
+                flat_args = flat_args[:]
```
do you really need a list copy here?

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -4631,9 +4726,8 @@ def functional_call(named_params, named_buffers, *args, **kwargs):
-    num_params_buffers = len(named_params) + len(named_buffers)
     compiled_f = aot_function(
-        functional_call, num_params_buffers=num_params_buffers, *args, **kwargs
+        functional_call, num_params_buffers=len(named_params) + len(named_buffers), *args, **kwargs
```
What's going with these changes?

ezyang reviewed (approved):
Please check comments

ezyang commented:
You should give @bdhirsh a chance to properly review too

tugsbayasgalan committed:
Update on "Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) 

We keep the input mutation in the graph by wrapping the original callable in a wrapper function where in the end we add input.copy_(updated_input) call which is then traced via make_fx. Previously, this was only enabled for forward-only path but unconditionally disabled for joint graph. 

Another caveat is that when we are tracing through tensor subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code logic quite ugly for no obvious performance improvement. 

Most of the changes in this PR are mechanical and I didn't have to make any change to the partitioner. Previously forward/backward heavily relied on metadata field `num_mutated_inps` to figure out whether something is returned as extra output or not. But now since we keep some mutations in the graph, we need to propogate something similar to `num_mutated_inps - num_graph_handled_inps`. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) 

We keep the input mutation in the graph by wrapping the original callable in a wrapper function where in the end we add input.copy_(updated_input) call which is then traced via make_fx. Previously, this was only enabled for forward-only path but unconditionally disabled for joint graph. 

Another caveat is that when we are tracing through tensor subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code logic quite ugly for no obvious performance improvement. 

Most of the changes in this PR are mechanical and I didn't have to make any change to the partitioner. Previously forward/backward heavily relied on metadata field `num_mutated_inps` to figure out whether something is returned as extra output or not. But now since we keep some mutations in the graph, we need to propogate something similar to `num_mutated_inps - num_graph_handled_inps`. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) 

We keep the input mutation in the graph by wrapping the original callable in a wrapper function where in the end we add input.copy_(updated_input) call which is then traced via make_fx. Previously, this was only enabled for forward-only path but unconditionally disabled for joint graph. 

Another caveat is that when we are tracing through tensor subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code logic quite ugly for no obvious performance improvement. 

Most of the changes in this PR are mechanical and I didn't have to make any change to the partitioner. Previously forward/backward heavily relied on metadata field `num_mutated_inps` to figure out whether something is returned as extra output or not. But now since we keep some mutations in the graph, we need to propogate something similar to `num_mutated_inps - num_graph_handled_inps`. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen commented:
subscribing to see when this PR lands

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) 

We keep the input mutation in the graph by wrapping the original callable in a wrapper function where in the end we add input.copy_(updated_input) call which is then traced via make_fx. Previously, this was only enabled for forward-only path but unconditionally disabled for joint graph. 

Another caveat is that when we are tracing through tensor subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code logic quite ugly for no obvious performance improvement. 

Most of the changes in this PR are mechanical and I didn't have to make any change to the partitioner. Previously forward/backward heavily relied on metadata field `num_mutated_inps` to figure out whether something is returned as extra output or not. But now since we keep some mutations in the graph, we need to propogate something similar to `num_mutated_inps - num_graph_handled_inps`. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/compile_fx.py:
```diff
@@ -1083,7 +1083,9 @@ def fw_compiler_base(
             context = torch._guards.TracingContext.get()
             # See Note [User Outputs in the inductor graph]
             if context is not None and context.fw_metadata and not is_inference:
-                original_output_start_index = context.fw_metadata.num_mutated_inputs
```
yeah agreed - I'm going to write up a list of AOTAutograd related refactor issues sometime soon, and this fits into one of them: "TracingContext shouldn't hold onto of `ViewAndMutationMeta`, just the stuff inductor needs"

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1577,9 +1605,9 @@ class AOTConfig:
         # We only need to bother cloning mutated inputs that participate in autograd.
-        mutated_inp_idx = meta.mutated_inp_indices.index(idx)
+        mutated_inp_idx = meta.mutated_inp_runtime_indices.index(idx)
         if meta.input_info[idx].requires_grad and meta.input_info[idx].mutates_data:
```
Is this requires_grad check necessary anymore? I think this codepath can now be:
```
if meta.input_info[idx].mutates_data:
    return t.clone()
elf meta.input_info[idx].mutates_metadata:
    return t.view(t.shape)
```

bdhirsh replied:
Nvm I think it still is: because we're saying that if we have subclass input that does not require grad but gets mutated, we will still not keep it in the graph (but we don't want to clone() it here)

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1837,12 +1865,12 @@ def functionalized_f_helper(*args):
-                    assert inpt_new is not inpt_old
-                    if meta.input_info[i].mutations_hidden_from_autograd:
-                        with torch.no_grad(), torch.autograd._unsafe_preserve_version_counter(inpt_old):
+                    if not trace_joint or (trace_joint and not inpt_old.requires_grad and not meta.input_info[i].requires_grad):
```
hmm what's this extra check for?

tugsbayasgalan replied:
This is to check when the input is updated so that it would be a tensor with grad. For example:
```
x = torch.tensor(1, requires_grad=False)
y = torch.tensor(1, requires_grad=True)
x.mul_(y)
```

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3029,27 +3084,43 @@ def runtime_wrapper(*args):
+                num_graph_handled = runtime_metadata.num_mutated_graph_handled_indices
+                # We return graph handled inputs in CompiledFunction.forward as we marked them with mark_dirty
+                if num_graph_handled > 0:
+                    all_outs = all_outs[:-num_graph_handled]
```
Ah right. Can you add a comment here mentioning that `autograd.Function` **requires** us to return the mutated inputs as extra outputs to the `autograd.Function.forward`? I think that's the real reason we're forced to do that here.

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3506,6 +3577,7 @@ def metadata_fn(*primals):
         metadata_fn,
         keep_input_mutations=meta.keep_input_mutations,
         is_train=meta.is_train,
+        is_subclass=True,
```
A comment here explaining the situation would be great. The entire reason we're plumbing this is_subclass argument everywhere here is being it's painful to support input mutations on subclasses, and that info isn't easily available.

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3594,6 +3667,7 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Any], aot_config: AOTConfig,
                 + inner_meta.num_outputs
                 + inner_meta.num_intermediate_bases
                 + inner_meta.num_outputs_rng_offset
+                - inner_meta.num_mutated_graph_handled_indices
```
instead of doing `inner_meta.num_mutated_inputs - inner_meta.num_mutated_graph_handled_indices`, can't we just do `inner_meta. num_mutated_runtime_inps`?

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3914,27 +3996,31 @@ def backward(ctx, *flat_args):
+            num_graph_handled_inputs = CompiledFunction.metadata.num_mutated_graph_handled_indices
             expected_grad_outs = (
-                CompiledFunction.metadata.num_outputs + num_mutated_inps + num_intermediate_bases
+                CompiledFunction.metadata.num_outputs + num_mutated_inps + num_intermediate_bases - num_graph_handled_inputs
```
Same thing here, just do `inner_meta.num_mutated_runtime_inps` instead of subtracting stuff

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3914,27 +3996,31 @@ def backward(ctx, *flat_args):
             assert len(flat_args) == expected_grad_outs
             out_info = CompiledFunction.metadata.output_info
 
+            num_mutated_inps_returned = num_mutated_inps - num_graph_handled_inputs
```
same here (this can be `num_runtime_handled_inputs`?)

bdhirsh reviewed (commented):
Overall looks good! Left a few comments.

tugsbayasgalan committed:
Update on "Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) 

We keep the input mutation in the graph by wrapping the original callable in a wrapper function where in the end we add input.copy_(updated_input) call which is then traced via make_fx. Previously, this was only enabled for forward-only path but unconditionally disabled for joint graph. 

Another caveat is that when we are tracing through tensor subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code logic quite ugly for no obvious performance improvement. 

Most of the changes in this PR are mechanical and I didn't have to make any change to the partitioner. Previously forward/backward heavily relied on metadata field `num_mutated_inps` to figure out whether something is returned as extra output or not. But now since we keep some mutations in the graph, we need to propogate something similar to `num_mutated_inps - num_graph_handled_inps`. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) 

We keep the input mutation in the graph by wrapping the original callable in a wrapper function where in the end we add input.copy_(updated_input) call which is then traced via make_fx. Previously, this was only enabled for forward-only path but unconditionally disabled for joint graph. 

Another caveat is that when we are tracing through tensor subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code logic quite ugly for no obvious performance improvement. 

Most of the changes in this PR are mechanical and I didn't have to make any change to the partitioner. Previously forward/backward heavily relied on metadata field `num_mutated_inps` to figure out whether something is returned as extra output or not. But now since we keep some mutations in the graph, we need to propogate something similar to `num_mutated_inps - num_graph_handled_inps`. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) 

We keep the input mutation in the graph by wrapping the original callable in a wrapper function where in the end we add input.copy_(updated_input) call which is then traced via make_fx. Previously, this was only enabled for forward-only path but unconditionally disabled for joint graph. 

Another caveat is that when we are tracing through tensor subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code logic quite ugly for no obvious performance improvement. 

Most of the changes in this PR are mechanical and I didn't have to make any change to the partitioner. Previously forward/backward heavily relied on metadata field `num_mutated_inps` to figure out whether something is returned as extra output or not. But now since we keep some mutations in the graph, we need to propogate something similar to `num_mutated_inps - num_graph_handled_inps`. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) 

We keep the input mutation in the graph by wrapping the original callable in a wrapper function where in the end we add input.copy_(updated_input) call which is then traced via make_fx. Previously, this was only enabled for forward-only path but unconditionally disabled for joint graph. 

Another caveat is that when we are tracing through tensor subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code logic quite ugly for no obvious performance improvement. 

Most of the changes in this PR are mechanical and I didn't have to make any change to the partitioner. Previously forward/backward heavily relied on metadata field `num_mutated_inps` to figure out whether something is returned as extra output or not. But now since we keep some mutations in the graph, we need to propogate something similar to `num_mutated_inps - num_graph_handled_inps`. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

bdhirsh reviewed (approved):
looks good :) - avoiding keeping input mutations in the graph when they are used in inference + require grad is a good followup

tugsbayasgalan committed:
Update on "Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) 

We keep the input mutation in the graph by wrapping the original callable in a wrapper function where in the end we add input.copy_(updated_input) call which is then traced via make_fx. Previously, this was only enabled for forward-only path but unconditionally disabled for joint graph. 

Another caveat is that when we are tracing through tensor subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code logic quite ugly for no obvious performance improvement. 

Most of the changes in this PR are mechanical and I didn't have to make any change to the partitioner. Previously forward/backward heavily relied on metadata field `num_mutated_inps` to figure out whether something is returned as extra output or not. But now since we keep some mutations in the graph, we need to propogate something similar to `num_mutated_inps - num_graph_handled_inps`. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Persist copy_ in training graph for inputs that don't require grad"


In this PR, we try to keep the input mutations in the forward graph IFF input mutation is data mutation and not metadata mutation and doesn't require grad. This is for optimizing inductor training graphs. (For more details: https://github.com/pytorch/pytorch/issues/109240) 

We keep the input mutation in the graph by wrapping the original callable in a wrapper function where in the end we add input.copy_(updated_input) call which is then traced via make_fx. Previously, this was only enabled for forward-only path but unconditionally disabled for joint graph. 

Another caveat is that when we are tracing through tensor subclasses, we won't allow any input mutations to be preserved in the graph. The reason is that it makes the code logic quite ugly for no obvious performance improvement. 

Most of the changes in this PR are mechanical and I didn't have to make any change to the partitioner. Previously forward/backward heavily relied on metadata field `num_mutated_inps` to figure out whether something is returned as extra output or not. But now since we keep some mutations in the graph, we need to propogate something similar to `num_mutated_inps - num_graph_handled_inps`. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge
